init/devices.cpp

```diff
@@ -458,7 +471,65 @@ void DeviceHandler::HandleUevent(const Uevent& uevent) {
 
     mkdir_recursive(Dirname(devpath), 0755);
 
-    HandleDevice(uevent.action, devpath, block, uevent.major, uevent.minor, links);
+    HandleDevice(uevent.action, devpath, uevent.path, block, uevent.major, uevent.minor, links);
+}
+
+bool DeviceHandler::GetDeviceAlias(const std::string &upath, int major, int minor,
+                                   std::string& alias_link) const {
+    bool found = false;
+    std::string parent_dir;
+    std::string vendorId_s, productId_s, dev_s;
+    std::string vendorId_path, productId_path;
+    std::string sAliasPath;
+    int vendorId, productId;
+
+    parent_dir = Dirname("/sys" + upath);
+
+    while ((parent_dir != "/" && parent_dir != ".")) {
+        std::string vendorId_path = parent_dir + "/idVendor";
+
+        // Get the vendor ID
+        if (!android::base::ReadFileToString(vendorId_path, &vendorId_s)) {
+            // Can't find the vendor ID? move up the path.
+            auto last_slash = parent_dir.rfind('/');
+            if (last_slash == std::string::npos) break;
+
+            parent_dir.erase(last_slash);
+        }
+        // Found the idVendor, move to the next step.
+        else { found = true; break; }
+    }
+
+    if (!found) return false;
+
+    for (const auto& alias : aliases_) {
+        bool match_productId, match_vendorId, match_major, match_minor;
+        std::string sys_path;
+
+        // Look in /sys for the information we need.
+        productId_path = parent_dir + "/idProduct";
+        sys_path = "/sys" + upath;
+
+        // Get the product ID
+        if (!android::base::ReadFileToString(productId_path, &productId_s)) {
+            PLOG(ERROR) << "Failed to read " << productId_path;
+            continue;
+        }
+
+        vendorId = std::stoi(vendorId_s, 0, 16);
+        productId = std::stoi(productId_s, 0, 16);
+        match_productId = (alias.ProductId() == productId);
+        match_vendorId = (alias.VendorId() == vendorId);
+        match_major = (alias.Major() == major);
+        match_minor = (alias.Minor() == minor);
+
+        if (match_productId && match_vendorId && match_major && match_minor) {
+            alias_link = alias.AliasTo();
+            return true;
```

Comment:
Does one alias for a given (productId, vendorId, major, minor) tuple covers the case of the modem which exposes a number of tty?

I haven't looked at this in details.